pkg/controller/secret/secret_controller.go

```diff
@@ -0,0 +1,428 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	alertmanager "github.com/openshift/configure-alertmanager-operator/pkg/types"
+
+	yaml "gopkg.in/yaml.v2"
+)
+
+var log = logf.Log.WithName("controller_secret")
+
+// Add creates a new Secret Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileSecret{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("secret-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource (type "Secret").
+	// For each Add/Update/Delete event, the reconcile loop will be sent a reconcile Request.
+	err = c.Watch(&source.Kind{Type: &corev1.Secret{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileSecret{}
+
+// ReconcileSecret reconciles a Secret object
+type ReconcileSecret struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a Secret object and makes changes based on the state read.
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileSecret) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling Secret")
+
+	// This operator is only interested in the 3 secrets listed below. Skip reconciling for all other secrets.
+	if request.Name != "alertmanager-main" && request.Name != "dms-secret" && request.Name != "pd-secret" {
+		reqLogger.Info("Skip reconcile: No changes detected to alertmanager secrets.")
+		return reconcile.Result{}, nil
+	}
+	fmt.Println("DEBUG: Started reconcile loop")
+
+	// This block looks at the specific instance of Secret for each Secret in the `openshift-monitoring` namespace.
```

Comment:
Do we want to monitor secrets on the `openshift-monitoring` ns as stated in that comment?

I noticed in the operator Deployment (06_operator.yaml file) that the WATCH_NAMESPACE parameter is set to the operator's ns (metadata.namespace), so the way it is it will watch for secrets only on the `configure-alertmanager-operator` ns

Reply:
Good catch! WATCH_NAMESPACE in this file is misleading and has no effect, because it's actually set here instead:

https://github.com/dak1n1/configure-alertmanager-operator/blob/initial_create/cmd/manager/main.go#L79

Should I configure that differently? For example, in the dedicated-admin-operator, the value `namespace` is used there instead:

https://github.com/openshift/dedicated-admin-operator/blob/master/cmd/manager/main.go#L89

But if I do that, it'll watch the namespace `configure-alertmanager-operator` because my operator pods run there. I think I'd better move my operator to run completely inside the `openshift-monitoring` namespace instead. That way I can configure mine like the dedicated-admin-operator. Does that sound ok?

Reply:
Sounds fine to me to move to `openshift-monitoring`.  Think we talked about this possibility a while back.  And if the secrets are synced to that NS as well then this operator is confined to a single NS.

Reply:
Ok, it's working in the openshift-monitoring namespace now. Thanks!